src/test/scala/parsec/HListParsersSpec.scala

```diff
@@ -0,0 +1,136 @@
+       with HListParsers
+       with CharParsers {
+
+  // Easier to type
+  type P[T] = Parser[T]
+
+  import shapeless._
+  import ops.hlist._
+
+  val reader = CharReader("oh3hiagain!".toArray)
+
+  test("parsing a single element gives a single result") {
+
+    val parsers = letter :: HNil
+    type Res = Char :: HNil
+    val res: ParseResult[Res] = 
+      parse[P[Char] :: HNil, True :: HNil, Res](parsers, reader)
+
+    checkSuccessH(res)(expected = 'o' :: HNil, expectedPos = 1)
+
+  }
+
+  test("parsing four elements works too") {
+
+    type Parsers = P[Char] :: P[Char] :: P[Char] :: P[Char] :: HNil
+    /* One less element here because of our projection! */
+    type Res = Char :: Char :: Char :: HNil
+    val parsers = letter :: letter :: digit :: letter :: HNil
+    val res: ParseResult[Res] = 
+      parse[Parsers, True :: False :: True :: True :: HNil, Res](parsers, reader)
+
+    checkSuccessH(res)(
+      expected = 'o' :: '3' ::'h' :: HNil,
+      expectedPos = 4
+    )
+
+  }
+
+  import util.Mappable._
+
+  test("concating three repped parsers works") {
+
+    val vowels = Set('a', 'e', 'i', 'o', 'u')
+    def isVowel(c: Char) = vowels contains c
+
+    type Res = String :: Int :: String :: HNil
+    type Projection = True :: True :: True :: HNil
+    type Parsers = Parser[String] :: Parser[Int] :: Parser[String] :: HNil
+    /* Represent `FoldParser` of each record in the parsed text */
+    val repConcats = letters :: rep(digit2Int) :: letters :: HNil
+
+    /* TODO Give a better type and don't expose the `FoldParser` */
+    def justTheStringPlease(fp: FoldParser[Char]): Parser[String] =
+      fp.toStringParser
+
+    def addtheInts(fp: FoldParser[Int]): Parser[Int] = 
+      fp.fold[Int](0, (acc, x) => acc + x)
+
+    def upperCaseTheVowels(fp: FoldParser[Char]): Parser[String] =
+      (fp filter isVowel map (_.toUpper)).toStringParser
+
+    /* TODO Have a better syntax here? */
+    val fs = (justTheStringPlease _) :: (addtheInts _) :: 
+             (upperCaseTheVowels _)  :: HNil
+
+    val parser: Parsers = map(repConcats, fs)
+    val input = CharReader("oh12345idris!".toArray)
+    val res: ParseResult[Res] =
+      parse[Parsers, Projection, Res](parser, input)
+
+    checkSuccessH(res)(
+      expected = "oh" :: 15 :: "II" :: HNil,
+      expectedPos = 12
+    )
+
+  }
+
+  test("parse a simple csv file (all elements are known)") {
+
+    import scala.collection.mutable.StringBuilder
+
+    val comma = accept(',')
+    val CRLF = accept('\n')
+
+    // Def because we need a new `StringBuilder` each time
+    def first: Parser[String] = letters.toStringParser
+    def last: Parser[String] = letters.toStringParser
+    def age: Parser[Int] = number
+
+    type Person = String :: String :: Int :: HNil
+    type Projection = True :: True :: True :: HNil
+    type PersonParser = Parser[String] :: 
+      Parser[String] :: Parser[Int] :: HNil
+
+    def personParser = mkParser[PersonParser, Projection, Person](
```

Comment:
same question here, do we need to provide the type parameters for `mkParser` explicitly?